runtime/tchannel_server.go

```diff
@@ -64,197 +83,339 @@ func (ncr netContextRouter) Handle(ctx netContext.Context, call *tchannel.Inboun
+	)
+	scope := s.scope.Tagged(map[string]string{
+		"endpoint": endpointID,
+		"handler":  handlerID,
+		"method":   method,
+	})
+	endpoint := TChannelEndpoint{
+		EndpointID:     endpointID,
+		HandlerID:      handlerID,
+		Method:         method,
+		handler:        h,
+		postResponseCB: cb,
+		Logger:         logger,
+		metrics:        NewInboundTChannelMetrics(scope),
 	}
-	s.register(service, method, handler)
+	s.register(&endpoint)
+	return &endpoint
 }
 
-func (s *TChannelRouter) register(service string, method string, h *handler) {
-	key := service + "::" + method
-
+func (s *TChannelRouter) register(e *TChannelEndpoint) {
 	s.Lock()
-	s.handlers[key] = *h
+	s.endpoints[e.Method] = e
 	s.Unlock()
 
 	ncr := netContextRouter{router: s}
-	s.registrar.Register(ncr, key)
+	s.registrar.Register(ncr, e.Method)
 }
 
 // Handle handles an incoming TChannel call and forwards it to the correct handler.
 func (s *TChannelRouter) Handle(ctx context.Context, call *tchannel.InboundCall) {
-	op := call.MethodString()
-	service, method, ok := getServiceMethod(op)
-	if !ok {
-		s.logger.Error(fmt.Sprintf("Handle got call for %s which does not match the expected call format", op))
+	method := call.MethodString()
+	if _, _, ok := getServiceMethod(method); !ok {
+		s.logger.Error(fmt.Sprintf("Handle got call for %s which does not match the expected call format", method))
+		return
 	}
 
 	s.RLock()
-	handler, ok := s.handlers[op]
+	e, ok := s.endpoints[method]
 	s.RUnlock()
 	if !ok {
-		s.logger.Error(fmt.Sprintf("Handle got call for %s which is not registered", op))
+		s.logger.Error(fmt.Sprintf("Handle got call for %s which is not registered", method))
+		return
 	}
 
-	if err := s.handle(ctx, handler, service, method, call); err != nil {
+	if err := s.handle(ctx, e, call); err != nil {
 		s.onError(err)
 	}
 }
 
+func getServiceMethod(method string) (string, string, bool) {
+	s := string(method)
+	sep := strings.Index(s, "::")
+	if sep == -1 {
+		return "", "", false
+	}
+	return s[:sep], s[sep+2:], true
+}
+
 func (s *TChannelRouter) onError(err error) {
 	if tchannel.GetSystemErrorCode(err) == tchannel.ErrCodeTimeout {
-		s.logger.Warn("Thrift server timeout",
-			zap.String("error", err.Error()),
-		)
+		s.logger.Warn("Thrift server timeout", zap.Error(err))
 	} else {
-		s.logger.Error("Thrift server error.",
-			zap.String("error", err.Error()),
-		)
+		s.logger.Error("Thrift server error.", zap.Error(err))
 	}
 }
 
-func (s *TChannelRouter) handle(ctx context.Context, handler handler, service string, method string, call *tchannel.InboundCall) error {
-	treader, err := call.Arg2Reader()
-	if err != nil {
-		return errors.Wrapf(err, "could not create arg2reader for inbound call: %s::%s", service, method)
+func (s *TChannelRouter) handle(
+	ctx context.Context,
+	e *TChannelEndpoint,
+	call *tchannel.InboundCall,
+) (err error) {
+	c := tchannelInboundCall{
+		endpoint: e,
+		call:     call,
 	}
+	defer func() { c.finish(err) }()
```

Comment:
this is not correct really. we do not always write to the body in this method.

For example we can `return err` and have the caller write an error frame on our behalf

So `finish(err)` happesn in this method only if we call `c.writeResBody()` otherwise `finish(err)` should be called upstream in the caller of this method.

Reply:
I'm not sure what you mean?  The `defer` function is always called where ever the `handle()` function returns, on any of the return statements. So `start()` and `finish()` are always called at the beginning and end of the `handle()` function, regardless of `c.writeResBody()` succeeding or not.